upcloud/error_utils.go

```diff
@@ -0,0 +1,87 @@
+package upcloud
+
+import (
+	"net/http"
+	"strings"
+)
+
+// IsNotFoundError checks if provided error is API Not Found error
+func IsNotFoundError(err error) bool {
+	if err == nil {
+		return false
+	}
+
+	if uError, ok := err.(*Error); ok {
+		return uError.Status == http.StatusNotFound
+	}
+
+	if uProblem, ok := err.(*Problem); ok {
+		return uProblem.Status == http.StatusNotFound
+	}
+
+	return false
+}
+
+// List of all error codes that indicate that API object already exists
+var alreadyExistsErrCodes = map[string]struct{}{
+	ErrCodeFirewallRuleExists: {},
+	ErrCodeDBExists:           {},
+	ErrCodeServiceExists:      {},
+	ErrCodeInterfaceExists:    {},
+	ErrCodeTagExists:          {},
+}
+
+// List of all json+problem type fragments that indicate that API object already exists
+var alreadyExistsJsonProblemTypes = map[string]struct{}{
+	ProblemTypeResourceAlreadyExists: {},
+}
+
+// IsAlreadyExistsError checks if provided error is API Already Exists error
+func IsAlreadyExistsError(err error) bool {
+	if err == nil {
+		return false
+	}
+
+	if uError, ok := err.(*Error); ok {
+		_, errCodeOk := alreadyExistsErrCodes[uError.ErrorCode]
+		return uError.Status == http.StatusConflict && errCodeOk
+	}
+
+	if uProblem, ok := err.(*Problem); ok {
+		problemType := GetJsonProblemType(uProblem)
+		_, problemTypeOk := alreadyExistsJsonProblemTypes[problemType]
+		return uProblem.Status == http.StatusBadRequest && problemTypeOk
+	}
+
+	return false
+}
+
+func IsAuthenticationFailedError(err error) bool {
+	if err == nil {
+		return false
+	}
+
+	if uError, ok := err.(*Error); ok {
+		return uError.Status == http.StatusUnauthorized && uError.ErrorCode == ErrCodeAuthenticationFailed
```

Comment:
Is there other kind of "authentication failed" errors than `http.StatusUnauthorized` ?

Reply:
I guess there aren't, let's just use status code here 👍